src/Phan/Language/Internal/FunctionSignatureMap.php

```
@@ -8641,7 +8642,7 @@
 'prev' => ['mixed', '&rw_array_arg'=>'array'],
 'print' => ['int', 'arg'=>'string'],
 'print_r' => ['string|true', 'var'=>'mixed', 'return='=>'bool'],
-'printf' => ['int', 'format'=>'string', '...args='=>'string|int|float'],
+'printf' => ['int', 'format'=>'string', '...args='=>'string|int|float|null'],
```

I'd consider it a bug if you deliberately called `$x = null; printf("%s", $x);`. There aren't any format specifiers I'm aware of that expect null. Users can pass `printf("%s", $x ?? '')` if they want to be clear about what it should do.

This is also causing a test failure in tests/plugin_test

---

I understand that the `null` is silently converted to `''`. There's a similar side-effect - `strict_types=1` disallows passing objects with __toString somewhere `string` is accepted, but `sprintf` accepts it and it works...

---

I think it's too opinionated to report this as an error, because it works in PHP: https://3v4l.org/KB4rC

I usually solve this by making the feature work in `phpstan/phpstan` but then writing a rule for `phpstan/phpstan-strict-rules` for those who want it to be reported.

---

> I think it's too opinionated to report this as an error, because it works in PHP: 

Which is fine for phpstan. Static analyzers can and already do have slightly different versions of the call maps (based on how their static analyzer is implemented, what philosophy they have on false positives/negatives, and what plugins they have) and just sync the most recent commits.

---

Alright 😊

---

> There's a similar side-effect - `strict_types=1` disallows passing objects with `__toString` somewhere `string` is accepted, but `sprintf` accepts it and it works...

Phan also implements the same check when strict_types is not set to 1, allowing objects with __toString() to be passed to printf specifiers expecting string, etc.

I reject `null` for reasons similar to rejecting `false`/`true` (they silently cast to 1 and the empty string, but are likely to be a mistake)